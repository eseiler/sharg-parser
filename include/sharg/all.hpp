// SPDX-FileCopyrightText: 2006-2024, Knut Reinert & Freie Universität Berlin
// SPDX-FileCopyrightText: 2016-2024, Knut Reinert & MPI für molekulare Genetik
// SPDX-License-Identifier: BSD-3-Clause

/*!\file
 * \author Svenja Mehringer <svenja.mehringer AT fu-berlin.de>
 * \brief Meta-header for the \link parser Parser module \endlink.
 */

/*!\defgroup parser Parser
 * \brief The Parser Module
 *
 * # The Parser Class
 *
 * \copydetails sharg::parser
 *
 * # Parsing Command Line Arguments
 *
 * \copydetails sharg::parser::parse
 *
 * # Argument Validation
 *
 * The Sharg Parser offers a validation mechanism for (positional_)options
 * via callables. You can pass any functor that fulfils the sharg::validator
 * and takes the value passed to the add_(positional_)option function call as
 * a parameter. We provide some commonly used functor that might come in handy:
 *
 * - sharg::regex_validator
 * - sharg::value_list_validator
 * - sharg::arithmetic_range_validator
 * - sharg::input_file_validator
 * - sharg::output_file_validator
 * - sharg::input_directory_validator
 * - sharg::output_directory_validator
 */

// Groups will appear in order they are defined.

/*!\defgroup exceptions Exceptions
 * \brief The Exceptions Module
 */

/*!\defgroup validators Validators
 * \brief The Validators Module
 */

/*!\defgroup misc Misc
 * \brief The Misc Module
 */

#pragma once

#include <sharg/auxiliary.hpp>
#include <sharg/exceptions.hpp>
#include <sharg/parser.hpp>
#include <sharg/validators.hpp>
