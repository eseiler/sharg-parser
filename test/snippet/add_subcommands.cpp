// SPDX-FileCopyrightText: 2006-2025 Knut Reinert & Freie Universität Berlin
// SPDX-FileCopyrightText: 2016-2025 Knut Reinert & MPI für molekulare Genetik
// SPDX-License-Identifier: CC0-1.0

#include <sharg/all.hpp>

void run(std::vector<std::string> const & arguments)
{
    sharg::parser git_parser{"git", arguments, sharg::update_notifications::off, {"pull", "push"}};
    git_parser.add_subcommands({"remote"});
    git_parser.parse();

    sharg::parser & sub_parser = git_parser.get_sub_parser();

    if (sub_parser.info.app_name == std::string_view{"git-pull"})
    {
        auto & pull_parser = sub_parser;
        std::string repository{};
        pull_parser.add_positional_option(repository, sharg::config{});
        pull_parser.parse();
    }
    else if (sub_parser.info.app_name == std::string_view{"git-push"})
    {
        auto & push_parser = sub_parser;
        std::string repository{};
        push_parser.add_positional_option(repository, sharg::config{});
        push_parser.parse();
    }
    else if (sub_parser.info.app_name == std::string_view{"git-remote"})
    {
        auto & remote_parser = sub_parser;
        remote_parser.add_subcommands({"set-url", "show"});
        remote_parser.parse();

        sharg::parser & recursive_sub_parser = remote_parser.get_sub_parser();

        if (recursive_sub_parser.info.app_name == std::string_view{"git-remote-set-url"})
        {
            auto & set_url_parser = recursive_sub_parser;
            std::string repository{};
            set_url_parser.add_positional_option(repository, sharg::config{});
            set_url_parser.parse();
        }
        else if (recursive_sub_parser.info.app_name == std::string_view{"git-remote-show"})
        {
            auto & show_parser = recursive_sub_parser;
            show_parser.parse();
        }
    }
}

int main(int argc, char ** argv)
{
    try
    {
        run({argv, argv + argc});
    }
    catch (sharg::parser_error const & ext)
    {
        std::cerr << "[Error] " << ext.what() << '\n';
        std::exit(-1);
    }

    return 0;
}
