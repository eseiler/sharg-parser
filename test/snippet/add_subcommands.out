git
===
    Try -h or --help for more information.
